### ISSUE TITLE
GET Request parameter double decoded

### ISSUE THREAD

**Issue by thagul** [grails/grails-core#10936]
I've created a fresh rest-api application with a controller containing a `show` action mapped at `GET /test/$id`. When the `id` parameter contains a `+` sign it's correctly decoded as a space but if the `+` is encoded as `%2B` in the url, the param is double decoded to a space while I would expect it to be a `+`
For example, I've a controller with just:

```
class TestController {
    static responseFormats = ['json', 'xml']
    
    def show() {
        log.info "Showing ID ${params.id}"
        respond id: params.id
    }
}

```
mapped to url `/test/$id`

If I open URL `http://localhost:8080/test/1+1` the response is `{"id":"1 1"}`
If I open URL `http://localhost:8080/test/1%2B1` the response is still `{"id":"1 1"}` while I expect it to be `{"id":"1+1"}`

Also in the log and debugging the application `params.id` is equal to `1 1` and not `1+1`

### Task List

- [x] Steps to reproduce provided
- [ ] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. create a fresh app with profile rest-api (grails create-app param-decoding --profile=rest-api)
2. create a new controller (grails create-controller Test)
3. create a `show` action as the one above
4. invoke the show action opening the url http://localhost:8080/test/1%2B1

### Expected Behaviour
`params.id` should be `1+1`

### Actual Behaviour
`params.id` is `1 1`

### Environment Information

- **Operating System**: macOS 10.13.3
- **Grails Version:** 3.3.2
- **JDK Version:** 1.8.0_131

### Example Application
https://github.com/thagul/param-decoding

**Comment by ilopmar** [comment]
According to https://stackoverflow.com/questions/5366007/why-does-the-encodings-of-a-url-and-the-query-string-part-differ/5433216#5433216 URIs and query params are encoded differently. 

RFC 3986 (https://www.rfc-editor.org/rfc/rfc3986.txt) explains that _The path is terminated by the first question mark ("?") or number sign ("#") character, or by the end of the URI._

> 3.3.  Path
>
>   The path component contains data, usually organized in hierarchical
>   form, that, along with data in the non-hierarchical query component
>   (Section 3.4), serves to identify a resource within the scope of the
>   URI's scheme and naming authority (if any).  The path is terminated
>   by the first question mark ("?") or number sign ("#") character, or
>   by the end of the URI.

So in your example `http://localhost:8080/test/1+1`, `1+1` is a valid path and the plus (`+`) sign is valid and shouldn't be decoded as `{"id":"1 1"}`.

The issue in the code is here: https://github.com/grails/grails-core/blob/3.3.x/grails-web-url-mappings/src/main/groovy/org/grails/web/mapping/AbstractUrlMappingInfo.java#L103

Because we're using `URLDecoder` instead of an Uri one. I've tried with `org.springframework.web.util.UriUtils.UriUtils.decode(id, encoding)` and now the result for both requests `http://localhost:8080/test/1%2B1` and `http://localhost:8080/test/1+1` is the same and is `{"id":"1+1"}`, which I think it's correct.

I personally think that we should change the implementation to use Spring's `UriUtils` but that particular line of code is there since 2009 so I'm pretty sure that this will be a breaking change.

@graemerocher, what do you think? Should I fix this in `master` instead of `3.3.x` branch?

**Comment by graemerocher** [comment]
master branch better yes

**Comment by ilopmar** [comment]
Submitted PR to `master` branch

**Comment by RonaldvB** [comment]
We still see this issue in on the Grails 4 branch, is there a work-around of fix available?

**Comment by davydotcom** [comment]
Fix will be in 4.1.x
